crates/wasi/src/tcp.rs

```diff
@@ -853,25 +898,103 @@ impl HostOutputStream for TcpWriteStream {
         }
         Ok(SOCKET_READY_SIZE)
     }
+
+    fn shutdown(&mut self) {
+        self.state = match mem::replace(&mut self.state, WriteState::Closed) {
+            // No write in progress, immediately shut down:
+            WriteState::Ready => {
+                native_shutdown(&self.stream, Shutdown::Write);
+                WriteState::Closed
+            }
+
+            // Schedule the shutdown after the current write has finished:
+            WriteState::Writing(write) => {
+                let stream = self.stream.clone();
+                WriteState::Closing(crate::runtime::spawn(async move {
+                    let result = write.await;
+                    native_shutdown(&stream, Shutdown::Write);
+                    result
+                }))
+            }
+
+            s => s,
+        };
+    }
+
     async fn cancel(&mut self) {
-        match mem::replace(&mut self.last_write, LastWrite::Closed) {
-            LastWrite::Waiting(task) => _ = task.abort_wait().await,
+        match mem::replace(&mut self.state, WriteState::Closed) {
+            WriteState::Writing(task) | WriteState::Closing(task) => _ = task.cancel().await,
             _ => {}
         }
     }
-}
 
-#[async_trait::async_trait]
-impl Subscribe for TcpWriteStream {
     async fn ready(&mut self) {
-        if let LastWrite::Waiting(task) = &mut self.last_write {
-            self.last_write = match task.await {
-                Ok(()) => LastWrite::Done,
-                Err(e) => LastWrite::Error(e),
-            };
+        match &mut self.state {
+            WriteState::Writing(task) => {
+                self.state = match task.await {
+                    Ok(()) => WriteState::Ready,
+                    Err(e) => WriteState::Error(e),
+                }
+            }
+            WriteState::Closing(task) => {
+                self.state = match task.await {
+                    Ok(()) => WriteState::Closed,
+                    Err(e) => WriteState::Error(e),
+                }
+            }
+            _ => {}
         }
-        if let LastWrite::Done = self.last_write {
+
+        if let WriteState::Ready = self.state {
             self.stream.writable().await.unwrap();
         }
     }
 }
+
+struct TcpWriteStream(Arc<Mutex<TcpWriter>>);
+
+#[async_trait::async_trait]
+impl HostOutputStream for TcpWriteStream {
+    fn write(&mut self, bytes: bytes::Bytes) -> Result<(), StreamError> {
+        try_lock_for_stream(&self.0)?.write(bytes)
+    }
+
+    fn flush(&mut self) -> Result<(), StreamError> {
+        try_lock_for_stream(&self.0)?.flush()
+    }
+
+    fn check_write(&mut self) -> Result<usize, StreamError> {
+        try_lock_for_stream(&self.0)?.check_write()
+    }
+
+    async fn cancel(&mut self) {
+        self.0.lock().await.cancel().await
+    }
+}
+
+#[async_trait::async_trait]
+impl Subscribe for TcpWriteStream {
+    async fn ready(&mut self) {
+        self.0.lock().await.ready().await
```

Comment:
Could this use the `try_lock_*` helpers instead of `lock()` to consistently trap on contention? That might help prevent covering up bugs by accident where some contention is handled in some places but not others.

Reply:
Subscribe::ready has no return type, so I can't return a trap

Reply:
Hm well actually as I look into looks like [stdio is already inconsistent](https://github.com/bytecodealliance/wasmtime/blob/a05baa387d4f0f34ea585ff37414a5646651c4c8/crates/wasi/src/stdio.rs#L106-L136), so let's leave this as-is.